ipv8/src/main/java/nl/tudelft/ipv8/messaging/eva/EVAProtocol.kt

```diff
@@ -0,0 +1,804 @@
+    private var community: Community,
+    scope: CoroutineScope,
+    var windowSizeInBlocks: Int = WINDOW_SIZE_IN_BLOCKS,
+    var retransmitIntervalInSec: Int = RETRANSMIT_INTERVAL_IN_SEC,
+    var retransmitAttemptCount: Int = RETRANSMIT_ATTEMPT_COUNT,
+    var scheduledSendIntervalInSec: Int = SCHEDULED_SEND_INTERVAL_IN_SEC,
+    var timeoutIntervalInSec: Int = TIMEOUT_INTERVAL_IN_SEC,
+    var binarySizeLimit: Int = BINARY_SIZE_LIMIT,
+    var terminateByTimeoutEnabled: Boolean = true,
+    var blockSize: Int = BLOCK_SIZE,
+    var retransmitEnabled: Boolean = true,
+    var loggingEnabled: Boolean = true
+) {
+    private var scheduled: MutableMap<Key, Queue<ScheduledTransfer>> = mutableMapOf()
+    private var incoming: MutableMap<Key, Transfer> = mutableMapOf()
+    private var outgoing: MutableMap<Key, Transfer> = mutableMapOf()
+    private var finishedIncoming: MutableMap<Key, MutableSet<String>> = mutableMapOf()
+    private var finishedOutgoing: MutableMap<Key, MutableSet<String>> = mutableMapOf()
+    private var scheduledTasks = PriorityQueue<ScheduledTask>()
+
+    var onReceiveProgressCallback: ((Peer, String, TransferProgress) -> Unit)? = null
+    var onReceiveCompleteCallback: ((Peer, String, String, ByteArray?) -> Unit)? = null
+    var onSendCompleteCallback: ((Peer, String, ULong) -> Unit)? = null
+    var onErrorCallback: ((Peer, TransferException) -> Unit)? = null
+
+    init {
+        if (loggingEnabled) logger.debug {
+            "EVAPROTOCOL: Initialized. Block size: ${blockSize}." +
+                "Window size: ${windowSizeInBlocks}." +
+                "Retransmit enabled: ${retransmitEnabled}." +
+                "Retransmit interval: ${retransmitIntervalInSec}sec." +
+                "Max retransmit attempts: ${retransmitAttemptCount}." +
+                "Terminate by timeout enabled: ${terminateByTimeoutEnabled}." +
+                "Timeout: ${timeoutIntervalInSec}sec." +
+                "Scheduled send interval: ${scheduledSendIntervalInSec}sec." +
+                "Binary size limit: ${binarySizeLimit}."
+        }
+
+        /**
+         * Coroutine that periodically executes the send scheduled transfers function
+         * with a send interval as the delay
+         */
+        scope.launch {
+            while (isActive) {
+                sendScheduled()
+
+                delay((scheduledSendIntervalInSec * 1000).toLong())
+            }
+        }
+
+        /**
+         * Coroutine that periodically checks whether there are scheduled tasks planned in the queue
+         * and if so, if its time to be executed. The interval is set to 1 second.
+         */
+        scope.launch {
+            while (isActive) {
+                while (scheduledTasks.any() && scheduledTasks.peek() != null && scheduledTasks.peek()!!.atTime <= Date().time) {
+                    val task = scheduledTasks.poll()!!
+
+                    task.action.invoke()
+                }
+
+                delay(1000L)
```

Comment:
Magic number. Perhaps place in a final somewhere on top.

Reply:
fixed in branch